presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java

```diff
@@ -66,7 +66,8 @@
     private final List<Integer> outputColumns;                        // elements are hive column indices
     private final Map<Integer, Type> columnTypes;                     // key: index into hiveColumnIndices array
     private final Object[] constantValues;                            // aligned with hiveColumnIndices array
-    private final List<FilterFunction> filterFunctions;
+    private final List<FilterFunction> filterFunctionsWithInputs;
```

Comment:
Separating filter functions into the ones with inputs and the one without is kinda artificial. What we actually need is a high-level ordering abstraction that can reorder filter functions in some way. For this specific implementation, it may not always be the case the the functions without input is selective or cheaper; for example `rand() > 0.9999999999`.

So in order to make it extensible in the future, how about having a `FilterOrdering` interface that takes a function list and returns an ordered function list? Then you could create a specific implementation for it that always put functions without input ahead of the one without. That interface can be extended with other implementation once we have profiling.

Reply:
Hi James! I've talked this over with @bhhari  and @oerling, and here's what I think our position is: 

It's definitely true that in the future, we will adaptively re-order instead of hardcoding this business logic. And I can see the case for laying more groundwork for that now. 

For now, though, we think we should not implement this suggestion, on the grounds of that's a big enough change that it requires its own pull request / commit. 

To give more color to that:
- This PR doesn't re-order FilterFunctions as much as chop them into two buckets -- those that happen *before* TupleDomainFilters and those that happen *after*. 
- If we wanted to re-order FilterFunctions *and* TupleDomainFilters, we'd have to come up with an interface that covers both of them, and re-structure the code to treat them the same. Right now they are quite different
- In reality, we aren't ordering filters or filterfunctions as much as columns -- so that adds more complexity to the proposed code

For these reasons (plus we are running against time to get this out quickly), we propose punting this proposed change to a different issue/PR.